x-attoparsec/src/X/Data/Attoparsec/Text.hs

```diff
@@ -31,8 +31,7 @@ positiveIntParser :: Parser Int
 positiveIntParser =
   (P.readMaybe <$> many1 digit) >>= P.maybe (fail "not a positive int") pure
 
--- | create a Parser a from a text Parser
-eitherText :: (T.Text -> Either T.Text a) -> Parser a
-eitherText p = do
-  t <- takeText
-  either fail pure (first T.unpack (p t))
+-- | create a Parser a from a text Parser and an input text
+--   usage: takeWhile (/= '-') >>= eitherText parseTimestamp
+eitherText :: (T.Text -> Either T.Text a) -> T.Text -> Parser a
+eitherText p = either (fail . T.unpack) pure . p
```

Comment:
That makes the syntax a bit different, using a '.'. You think it's more idiomatic to have a minimal definition and have clients use function composition?
```
-- | create a Parser a from a text Parser and an input text
--   usage: takeWhile (/= '-') >>= eitherText parseTimestamp
eitherText :: (T.Text -> Either T.Text a) -> T.Text -> Parser a
eitherText p = eitherText' . p

-- | create a Parser a from a parse result
--   usage: takeWhile (/= '-') >>= eitherText' . parseTimestamp
eitherText' :: Either T.Text a -> Parser a
eitherText' = either (fail . T.unpack) pure
```